gstreamer-pbutils/src/encoding_profile.rs

```diff
@@ -0,0 +1,578 @@
+        }
+    }
+
+    pub fn build(self) -> Result<EncodingProfile, EncodingProfileBuilderError> {
+        if self.base.format.is_none() {
+            return Err(EncodingProfileBuilderError);
+        }
+
+        let profile = EncodingAudioProfile::new(
+            self.base.format.unwrap(), self.base.preset,
+            self.base.restriction, self.base.presence);
+
+        let mut profile = profile.upcast();
+        set_common_fields(&mut profile, &self.base);
+        Ok(profile)
+    }
+}
+
+pub struct EncodingVideoProfileBuilder<'a> {
+    base : EncodingProfileBuilderData<'a>,
+    pass: u32,
+    variable_framerate: bool,
+}
+
+declare_encoding_profile_builder!(EncodingVideoProfileBuilder);
+
+impl<'a> EncodingVideoProfileBuilder<'a> {
+    pub fn new() -> Self {
+        EncodingVideoProfileBuilder {
+            base: EncodingProfileBuilderData::new(),
+            pass: 0,
+            variable_framerate: false
+        }
+    }
+
+    pub fn pass(mut self, pass: u32) -> Self {
+        self.pass = pass;
+        self
+    }
+
+    pub fn variable_framerate(mut self, variable_framerate: bool) -> Self {
+        self.variable_framerate = variable_framerate;
+        self
+    }
+
+    pub fn build(self) -> Result<EncodingProfile, EncodingProfileBuilderError> {
+        if self.base.format.is_none() {
+            return Err(EncodingProfileBuilderError);
+        }
+
+        let video_profile = EncodingVideoProfile::new(
+            self.base.format.unwrap(), self.base.preset, self.base.restriction, self.base.presence);
+
+        video_profile.set_pass(self.pass);
+        video_profile.set_variableframerate(self.variable_framerate);
+
+        let mut profile = video_profile.upcast();
+        set_common_fields(&mut profile, &self.base);
+        Ok(profile)
+    }
+}
+
+pub struct EncodingContainerProfileBuilder<'a> {
+    base : EncodingProfileBuilderData<'a>,
+    profiles: LinkedList<& 'a EncodingProfile>
+}
+
+declare_encoding_profile_builder!(EncodingContainerProfileBuilder);
+
+impl<'a> EncodingContainerProfileBuilder<'a> {
+    pub fn new() -> Self {
+        EncodingContainerProfileBuilder {
+            base: EncodingProfileBuilderData::new(),
+            profiles: LinkedList::new()
+        }
+    }
+
+    pub fn build(self) -> Result<EncodingProfile, EncodingProfileBuilderError> {
+        if self.base.format.is_none() {
+            return Err(EncodingProfileBuilderError);
+        }
+
+        let container_profile = EncodingContainerProfile::new(
+            self.base.name, self.base.description, self.base.format.unwrap(), self.base.preset);
+
+        for profile in self.profiles {
+            container_profile.add_profile(profile).or_else(|_error| Err(EncodingProfileBuilderError))?;
+        }
+
+        let mut profile = container_profile.upcast();
+        set_common_fields(&mut profile, &self.base);
+        Ok(profile)
+    }
+
+    pub fn add_profile(mut self, profile: & 'a EncodingProfile) -> Self {
```

Comment:
As mentioned above, make this generic over the `IsA` trait :) Makes life easier for users

Basically, in the end there should be no explicit casts anymore anywhere